network/dag/publisher.go

```diff
@@ -0,0 +1,89 @@
+package dag
+
+import (
+	"github.com/nuts-foundation/nuts-node/crypto/hash"
+	"github.com/nuts-foundation/nuts-node/network/log"
+)
+
+// NewReplayingDAGPublisher creates a DAG publisher that replays the complete DAG to all subscribers when started.
+func NewReplayingDAGPublisher(payloadStore PayloadStore, dag DAG) Publisher {
+	publisher := &replayingDAGPublisher{
+		subscribers:  map[string]Receiver{},
+		algo:         NewBFSWalkerAlgorithm().(*bfsWalkerAlgorithm),
+		payloadStore: payloadStore,
+		dag:          dag,
+	}
+	dag.RegisterObserver(publisher.DocumentAdded)
+	payloadStore.RegisterObserver(publisher.PayloadWritten)
+	return publisher
+}
+
+type replayingDAGPublisher struct {
+	subscribers  map[string]Receiver
+	algo         *bfsWalkerAlgorithm
+	payloadStore PayloadStore
+	dag          DAG
+}
+
+func (s *replayingDAGPublisher) PayloadWritten(_ interface{}) {
+	s.publish(hash.EmptyHash())
```

Comment:
why the empty hash?

Reply:
Because when the node receives payload, we'd have to look up in the `payload hash -> document` index where we should resume (and check whether that's a resumation point) but that's not accessible from this type (it's an internal of the BBolt DAG implementation). So I chose to just resume from the resumation points regardless whether we can actually go forward with publishing to subscribers (it can only resume when the payload's there)

When we receive payload chances are pretty big that its from a resumation point, especially when your node is in sync with other nodes. We could optimize this later if it turns out to be a bottleneck.